server/src/packages/core/index.ts

```diff
@@ -1,6 +1,38 @@
-import * as t from 'io-ts';
+import {string, refinement, TypeOf } from 'io-ts';
 
 export const uuidCheck = /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i;
-export const UUIDRefinement = t.refinement(t.string, (str: string) => uuidCheck.test(str), 'Uuid');
+export const UUIDRefinement = refinement(string, (str: string) => uuidCheck.test(str), 'Uuid');
 
-export type Uuid = t.TypeOf<typeof UUIDRefinement>;
+// tslint:disable:max-classes-per-file
+export type Uuid = TypeOf<typeof UUIDRefinement>;
+
+// All this is testing trying to make the UUID types more specific
+// https://github.com/Microsoft/TypeScript/issues/202#issuecomment-498201335
+export function branded<T, Brand>() {
+  return class BrandedType {
+      value: BrandedType;
+      '__ kind': Brand;
+      static toBranded<Cls extends typeof BrandedType>(this: Cls, t: T) {
+        return t as unknown as InstanceType<Cls>;
+      }
+      static fromBranded<Cls extends typeof BrandedType>(this: Cls, b: InstanceType<Cls>) {
+        return b as unknown as T;
+      }
+      static Type: BrandedType;
+  };
+}
+
+// Modified for UUID
+export function uuidType<Brand>() {
+  return class UuidType {
+      value: UuidType;
+      '__ kind': Brand;
+      static fromUuid<Cls extends typeof UuidType>(this: Cls, t: Uuid) {
```

Comment:
Would it be an idea to run the regex on the input string here?

Reply:
Ooo - yes!